src/ovs/aca_ovs_l3_programmer.cpp

```diff
@@ -42,49 +42,346 @@ ACA_OVS_L3_Programmer &ACA_OVS_L3_Programmer::get_instance()
+              routing_rule_table_entry.next_hop_mac =
+                      current_routing_rule.routing_rule_extra_info().next_hop_mac();
+
+              if (!is_routing_rule_exist) {
+                new_subnet_routing_table_entry.routing_rules.emplace(
+                        current_routing_rule.id(), routing_rule_table_entry);
+
+                ACA_LOG_INFO("Added routing table entry for routering rule id %s\n",
+                             current_routing_rule.id().c_str());
+              } else {
+                ACA_LOG_INFO("Using existing routing table entry for routering rule id %s\n",
+                             current_routing_rule.id().c_str());
+              }
+
+            } else if (current_routing_rule.operation_type() == OperationType::DELETE) {
+              if (new_subnet_routing_table_entry.routing_rules.erase(
+                          current_routing_rule.id())) {
+                ACA_LOG_INFO("Successfuly cleaned up entry for router rule id %s\n",
+                             current_routing_rule.id().c_str());
+              } else {
+                ACA_LOG_ERROR("Failed to clean up entry for router rule id %s\n",
+                              current_routing_rule.id().c_str());
+                overall_rc = EXIT_FAILURE;
+              }
+            } else {
+              ACA_LOG_ERROR("Invalid operation_type: %d for router_rule with ID: %s.\n",
+                            current_routing_rule.operation_type(),
+                            current_routing_rule.id().c_str());
+              overall_rc = EXIT_FAILURE;
+            }
+          }
+
+          if (!is_subnet_routing_table_exist) {
+            new_subnet_routing_tables.emplace(current_router_subnet_id,
+                                              new_subnet_routing_table_entry);
+
+            ACA_LOG_INFO("Added router subnet table entry for subnet id %s\n",
+                         current_router_subnet_id.c_str());
+          } else {
+            ACA_LOG_INFO("Using existing router subnet table entry for subnet id %s\n",
+                         current_router_subnet_id.c_str());
+          }
+
+          subnet_info_found = true;
+          break;
+        }
+
+        if (!subnet_info_found) {
+          ACA_LOG_ERROR("Not able to find the info for router with subnet ID: %s.\n",
+                        current_router_subnet_id.c_str());
+          overall_rc = -EXIT_FAILURE;
+        }
+      } // for (int j = 0; j < parsed_struct.subnet_states_size(); j++)
+    } // for (int i = 0; i < current_RouterConfiguration.subnet_routing_tables_size(); i++)
+
+    if (!is_router_exist ||
+        (current_RouterConfiguration.message_type() == MessageType::FULL)) {
+      // -----critical section starts-----
+      _routers_table_mutex.lock();
+      _routers_table.emplace(router_id, new_subnet_routing_tables);
+      _routers_table_mutex.unlock();
+      // -----critical section ends-----
```

Comment:
Please remember to create an issue for this part, to remind ourselves.

Reply:
NVM. Found a tracking issue https://github.com/futurewei-cloud/alcor-control-agent/issues/129 👍